repository/src/main/java/org/alfresco/repo/security/authentication/identityservice/IdentityServiceJITProvisioningHandler.java

```diff
@@ -0,0 +1,171 @@
+ * from {@link IdentityServiceFacade.DecodedAccessToken} or {@link UserInfo}
+ * and creates a new user if it does not exist in the repository.
+ */
+public class IdentityServiceJITProvisioningHandler
+{
+    private final IdentityServiceFacade identityServiceFacade;
+    private final PersonService personService;
+    private final TransactionService transactionService;
+
+    private final Function<IdentityServiceFacade.DecodedAccessToken, Optional<? extends OIDCUserInfo>> mapTokenToUserInfoResponse = token -> {
+        Optional<String> firstName = Optional.ofNullable(token.getClaim(PersonClaims.GIVEN_NAME_CLAIM_NAME))
+            .filter(String.class::isInstance)
+            .map(String.class::cast);
+        Optional<String> lastName = Optional.ofNullable(token.getClaim(PersonClaims.FAMILY_NAME_CLAIM_NAME))
+            .filter(String.class::isInstance)
+            .map(String.class::cast);
+        Optional<String> email = Optional.ofNullable(token.getClaim(PersonClaims.EMAIL_CLAIM_NAME))
+            .filter(String.class::isInstance)
+            .map(String.class::cast);
+
+        return Optional.ofNullable(token.getClaim(PersonClaims.PREFERRED_USERNAME_CLAIM_NAME))
+            .filter(String.class::isInstance)
+            .map(String.class::cast).map(this::normalizeUserId)
+            .map(username -> new OIDCUserInfo(username, firstName.orElse(""), lastName.orElse(""), email.orElse("")));
+    };
+
+    public IdentityServiceJITProvisioningHandler(IdentityServiceFacade identityServiceFacade,
+        PersonService personService,
+        TransactionService transactionService)
+    {
+        this.identityServiceFacade = identityServiceFacade;
+        this.personService = personService;
+        this.transactionService = transactionService;
+    }
+
+    public Optional<OIDCUserInfo> extractUserInfoAndCreateUserIfNeeded(String bearerToken)
+    {
+        Optional<OIDCUserInfo> userInfoResponse = Optional.ofNullable(bearerToken)
+            .filter(Predicate.not(String::isEmpty))
+            .flatMap(token -> extractUserInfoResponseFromAccessToken(token).flatMap(
+                response -> response.username() == null || response.username().isEmpty() ?
+                    extractUserInfoResponseFromEndpoint(token) :
+                    Optional.of(response)).or(() -> extractUserInfoResponseFromEndpoint(token)));
+
+        if (transactionService.isReadOnly() || userInfoResponse.isEmpty())
+        {
+            return userInfoResponse;
+        }
+        return AuthenticationUtil.runAs(new AuthenticationUtil.RunAsWork<Optional<OIDCUserInfo>>()
+        {
+            @Override
+            public Optional<OIDCUserInfo> doWork() throws Exception
+            {
+                return userInfoResponse.map(userInfo -> {
+                    if (userInfo.username() != null && !personService.personExists(userInfo.username())
+                        && personService.createMissingPeople())
```

Comment:
```suggestion
                    if (userInfo.username() != null && personService.createMissingPeople()
                        && !personService.personExists(userInfo.username()))
```

Wouldn't it be better to keep less expensive checks earlier in the chain so that we don't hit the database unnecessarily in case `createMissingPeople` is false anyways?